File: frontend/src/api/workouts.ts
import type { Workout } from "../workout/types";
import type { WorkoutSummary } from "../workout/summary-types";

export async function startWorkout(personId: number): Promise<Workout> {
  const response = await fetch(`/api/persons/${personId}/workouts`, {
    method: "POST",
  });

  if (!response.ok) {
    throw new Error(`Could not start workout: HTTP ${response.status}`);
  }

  return (await response.json()) as Workout;
}

export async function checkpointWorkout(
  workoutId: string,
  elapsedSeconds: number,
  distanceM: number,
  videoPositionSeconds: number,
): Promise<Workout> {
  const response = await fetch(`/api/workouts/${workoutId}/checkpoint`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      elapsedSeconds,
      distanceM,
      videoPositionSeconds,
    }),
  });

  if (!response.ok) {
    throw new Error(`Could not checkpoint workout: HTTP ${response.status}`);
  }

  return (await response.json()) as Workout;
}

export async function completeWorkout(
  workoutId: string,
  elapsedSeconds: number,
  distanceM: number,
): Promise<Workout> {
  const response = await fetch(`/api/workouts/${workoutId}/complete`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      elapsedSeconds,
      distanceM,
    }),
  });

  if (!response.ok) {
    throw new Error(`Could not complete workout: HTTP ${response.status}`);
  }

  return (await response.json()) as Workout;
}

export async function abortWorkout(
  workoutId: string,
  elapsedSeconds: number,
  distanceM: number,
): Promise<Workout> {
  const response = await fetch(`/api/workouts/${workoutId}/abort`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      elapsedSeconds,
      distanceM,
    }),
  });

  if (!response.ok) {
    throw new Error(`Could not abort workout: HTTP ${response.status}`);
  }

  return (await response.json()) as Workout;
}

export async function getWorkoutHistory(
  personId: number,
  limit = 20,
): Promise<Workout[]> {
  const response = await fetch(
    `/api/persons/${personId}/workouts?limit=${limit}`,
  );

  if (!response.ok) {
    throw new Error(`Could not load workout history: HTTP ${response.status}`);
  }

  return (await response.json()) as Workout[];
}

export async function getWorkoutSummary(
  workoutId: string,
): Promise<WorkoutSummary> {
  const response = await fetch(`/api/workouts/${workoutId}/summary`);

  if (!response.ok) {
    throw new Error(`Could not load workout summary: HTTP ${response.status}`);
  }

  return (await response.json()) as WorkoutSummary;
}

File: backend/apps/api/routers/workouts.py
from fastapi import APIRouter, HTTPException

from application.workout.service import (
    InvalidWorkoutDurationError,
    InvalidWorkoutVideoError,
    WorkoutAlreadyFinishedError,
    WorkoutNotFoundError,
)
from apps.api import wiring
from apps.api.recommendation import create_training_recommendation
from contracts.workout import (
    FinishWorkoutRequest,
    StartWorkoutRequest,
    WorkoutCheckpointRequest,
    WorkoutResponse,
    WorkoutSummaryResponse,
)
from contracts.workout_mapper import (
    to_workout_response,
    to_workout_summary_response,
)

router = APIRouter()


@router.get(
    "/api/persons/{person_id}/workouts",
    response_model=list[WorkoutResponse],
    response_model_by_alias=True,
)
async def workout_history(
    person_id: int,
    limit: int = 20,
) -> list[WorkoutResponse]:
    person = wiring.person_service.get_person(person_id)

    if person is None:
        raise HTTPException(
            status_code=404,
            detail="Person not found",
        )

    if limit < 1 or limit > 100:
        raise HTTPException(
            status_code=422,
            detail="Limit must be between 1 and 100",
        )

    workouts = wiring.workout_service.get_for_person(
        person_id,
        limit=limit,
    )

    return [to_workout_response(workout) for workout in workouts]


@router.post(
    "/api/persons/{person_id}/workouts",
    response_model=WorkoutResponse,
    response_model_by_alias=True,
)
async def start_workout(
    person_id: int,
    request: StartWorkoutRequest | None = None,
) -> WorkoutResponse:
    person = wiring.person_service.get_person(person_id)

    if person is None:
        raise HTTPException(
            status_code=404,
            detail="Person not found",
        )

    recommendation = create_training_recommendation(person_id)

    try:
        workout = wiring.workout_service.start(
            person_id=person_id,
            recommendation=recommendation,
            video_id=(request.video_id if request is not None else None),
        )
    except InvalidWorkoutVideoError as exc:
        raise HTTPException(
            status_code=422,
            detail=str(exc),
        ) from exc

    return to_workout_response(workout)


@router.post(
    "/api/workouts/{workout_id}/checkpoint",
    response_model=WorkoutResponse,
    response_model_by_alias=True,
)
async def checkpoint_workout(
    workout_id: str,
    request: WorkoutCheckpointRequest,
) -> WorkoutResponse:
    try:
        workout = wiring.workout_service.checkpoint(
            workout_id,
            elapsed_seconds=request.elapsed_seconds,
            distance_m=request.distance_m,
            video_position_seconds=(request.video_position_seconds),
        )
    except WorkoutNotFoundError as exc:
        raise HTTPException(
            status_code=404,
            detail=str(exc),
        ) from exc
    except WorkoutAlreadyFinishedError as exc:
        raise HTTPException(
            status_code=409,
            detail=str(exc),
        ) from exc
    except InvalidWorkoutDurationError as exc:
        raise HTTPException(
            status_code=422,
            detail=str(exc),
        ) from exc

    return to_workout_response(workout)


@router.post(
    "/api/workouts/{workout_id}/complete",
    response_model=WorkoutResponse,
    response_model_by_alias=True,
)
async def complete_workout(
    workout_id: str,
    request: FinishWorkoutRequest,
) -> WorkoutResponse:
    try:
        workout = wiring.workout_service.complete(
            workout_id,
            elapsed_seconds=request.elapsed_seconds,
            distance_m=request.distance_m,
        )
    except WorkoutNotFoundError as exc:
        raise HTTPException(
            status_code=404,
            detail=str(exc),
        ) from exc
    except WorkoutAlreadyFinishedError as exc:
        raise HTTPException(
            status_code=409,
            detail=str(exc),
        ) from exc
    except InvalidWorkoutDurationError as exc:
        raise HTTPException(
            status_code=422,
            detail=str(exc),
        ) from exc

    return to_workout_response(workout)


@router.post(
    "/api/workouts/{workout_id}/abort",
    response_model=WorkoutResponse,
    response_model_by_alias=True,
)
async def abort_workout(
    workout_id: str,
    request: FinishWorkoutRequest,
) -> WorkoutResponse:
    try:
        workout = wiring.workout_service.abort(
            workout_id,
            elapsed_seconds=request.elapsed_seconds,
            distance_m=request.distance_m,
        )
    except WorkoutNotFoundError as exc:
        raise HTTPException(
            status_code=404,
            detail=str(exc),
        ) from exc
    except WorkoutAlreadyFinishedError as exc:
        raise HTTPException(
            status_code=409,
            detail=str(exc),
        ) from exc
    except InvalidWorkoutDurationError as exc:
        raise HTTPException(
            status_code=422,
            detail=str(exc),
        ) from exc

    return to_workout_response(workout)


@router.get(
    "/api/workouts/{workout_id}/summary",
    response_model=WorkoutSummaryResponse,
    response_model_by_alias=True,
)
async def workout_summary(
    workout_id: str,
) -> WorkoutSummaryResponse:
    try:
        summary = wiring.workout_service.get_summary(workout_id)
    except WorkoutNotFoundError as exc:
        raise HTTPException(
            status_code=404,
            detail=str(exc),
        ) from exc

    return to_workout_summary_response(summary)
